tests/snmp/airos.py

```diff
@@ -3,38 +3,54 @@
 from netengine.exceptions import NetEngineError
 
 from ..settings import settings
+from ..static import MockOutputMixin
 
 
 __all__ = ['TestSNMPAirOS']
 
 
-class TestSNMPAirOS(unittest.TestCase):
+class TestSNMPAirOS(unittest.TestCase, MockOutputMixin):
     
     def setUp(self):
         self.host = settings['airos-snmp']['host']
         self.community = settings['airos-snmp']['community']
         self.port = settings['airos-snmp'].get('port', 161)
-        
         self.device = AirOS(self.host, self.community, port=self.port)
+
+        # mock calls being made to devices
+        self.oid_mock_data = self._load_mock_json('/test-airos-snmp.json')
+        self.nextcmd_patcher = self._patch(
+            'netengine.backends.snmp.base.cmdgen.CommandGenerator.nextCmd',
+            return_value=[0, 0, 0, [[[0, 1]]] * 5]
+        )
+        self.getcmd_patcher = self._patch(
+            'netengine.backends.snmp.base.cmdgen.CommandGenerator.getCmd',
+            side_effect=lambda *args: self._get_mocked_getcmd(
+                data=self.oid_mock_data, input=args
+            )
+        )
+        self.getcmd_patcher.start()
     
     def test_get_value_error(self):
-        with self.assertRaises(NetEngineError):
-            self.device.get_value('.')
+        with self.getcmd_patcher as p:
+            p.side_effect = NetEngineError
+            with self.assertRaises(NetEngineError):
+                self.device.get_value('.')
```

Comment:
It is not clear to me what this test is doing. Can you please explain. 

From what I understood, it is testing this code

https://github.com/openwisp/netengine/blob/4f409d85ce7d2e18a2463c6bd847cab7a7f67e47/netengine/backends/snmp/base.py#L100-L109

But, `self.getcmd_patcher` here is mocking `netengine.backends.snmp.base.cmdgen.CommandGenerator.getCmd` which is an external method and it can't raise `NetEngineError`. 

To test whether `NetEngineError` is raised, I would somehow raise `IndexError` in above code. 

Is my understanding correct? 

In the current test, if I replace  `self.device.get_value('.')` with a valid value then it should fail, right?

Reply:
@pandafy Yes there is a mistake here, we don't need to raise an exception manually. Since there is no "." in the JSON file we stored it'd raise an exception itself.

> In the current test, if I replace self.device.get_value('.') with a valid value then it should fail, right?

Yes but after doing what I said above ^, it works fine

Reply:
i added this test in `tests/snmp/openwrt.py` and it failed with this error:

```
/home/pandafy/openwisp/venv-python2/lib/python2.7/site-packages/paramiko/transport.py:33: CryptographyDeprecationWarning: Python 2 is no longer supported by the Python core team. Support for it is now deprecated in cryptography, and will be removed in the next release.
  from cryptography.hazmat.backends import default_backend
E
======================================================================
ERROR: test_get_value_error (tests.snmp.TestSNMPOpenWRT)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/pandafy/openwisp/netengine/tests/snmp/openwrt.py", line 21, in test_get_value_error
    self.device.get_value('.')
  File "/home/pandafy/openwisp/netengine/netengine/backends/snmp/base.py", line 105, in get_value
    result = self.get(oid)
  File "/home/pandafy/openwisp/netengine/netengine/backends/snmp/base.py", line 83, in get
    return self._command.getCmd(self.community, self.transport, self._oid(oid))
  File "/home/pandafy/openwisp/venv-python2/lib/python2.7/site-packages/pysnmp/entity/rfc3413/oneliner/cmdgen.py", line 183, in getCmd
    **kwargs):
  File "/home/pandafy/openwisp/venv-python2/lib/python2.7/site-packages/pysnmp/hlapi/asyncore/sync/cmdgen.py", line 111, in getCmd
    lookupMib=options.get('lookupMib', True)))
  File "/home/pandafy/openwisp/venv-python2/lib/python2.7/site-packages/pysnmp/hlapi/asyncore/cmdgen.py", line 130, in getCmd
    vbProcessor.makeVarBinds(snmpEngine, varBinds), __cbFun,
  File "/home/pandafy/openwisp/venv-python2/lib/python2.7/site-packages/pysnmp/hlapi/varbinds.py", line 39, in makeVarBinds
    __varBinds.append(varBind.resolveWithMib(mibViewController, ignoreErrors=False))
  File "/home/pandafy/openwisp/venv-python2/lib/python2.7/site-packages/pysnmp/smi/rfc1902.py", line 853, in resolveWithMib
    self.__args[0].resolveWithMib(mibViewController)
  File "/home/pandafy/openwisp/venv-python2/lib/python2.7/site-packages/pysnmp/smi/rfc1902.py", line 421, in resolveWithMib
    self.__oid
  File "/home/pandafy/openwisp/venv-python2/lib/python2.7/site-packages/pysnmp/smi/view.py", line 211, in getNodeNameByOid
    (modName, nodeName, self)
NoSuchObjectError: NoSuchObjectError({'str': "Can't resolve node name :: at <pysnmp.smi.view.MibViewController object at 0x7f425e825950>"})
-------------------- >> begin captured stdout << ---------------------
DEBUG: SNMP GET .

--------------------- >> end captured stdout << ----------------------

----------------------------------------------------------------------
Ran 1 test in 0.092s

FAILED (errors=1)
```

Maybe the internal working of `pysnmp` has evolved over time. You might need to tweak code too. I am not sure on this. 

cc @nemesisdesign